Handle optional dependencies correctly - Fixes #286

yabberyabber opened the pull request:
The previous behavior around optional dependencies was incorrect in
some cases. If plugin A declares a required dependency on plugin C
at version N, and plugin B declares an optional dependency on plugin
C at version N+1, the resolver will completely ignore plugin B's
optional dependency and end up installing plugin C at version N. This
will break plugin B because plugin B is explicitly incompatible with
versions vefore its dependency.

To solve this, we keep optional plugins in the dependency graph until
the very end. If every dependency on a given plugin is optional, then
that plugin can be omitted from the resulting set. However, if any of
those plugins are non-optional, then that plugin must be included
in the result and the version of that plugin must be the highest
required version among all enumerated dependencies.

timja reviewed (commented):
can you write some tests that cover this properly please

yabberyabber commented:
Added some tests and fixed most of the unit tests.

One of the existing unit tests is not passing and I'm not sure why - I'll update this PR with a fix to that failing test once I figure out the issue.

yabberyabber commented:
This approach has an issue - non-optional dependencies of optional dependencies are treated as non-optional dependencies even if those optional dependencies don't end up being mapped in. Will address on Monday and add additional test for this scenario

timja commented on plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java:
```diff
@@ -930,7 +956,7 @@ public VersionNumber getLatestPluginVersion(String pluginName) {
     // TODO(oleg_nenashev): This method is private, because it is only a partial fix for https://github.com/jenkinsci/plugin-installation-manager-tool/issues/101
     // A full dependency graph resolution and removal of non-needed dependency trees is required
-    private Map<String, Plugin> resolveRecursiveDependencies(Plugin plugin, @CheckForNull Map<String, Plugin> topLevelDependencies) {
+    public Map<String, Plugin> resolveRecursiveDependencies(Plugin plugin, @CheckForNull Map<String, Plugin> topLevelDependencies) {
```
there's a comment on line 957 that needs removing now

timja replied:
(removed)
